rfcs/rdfa-editor-stage-1.md

```diff
@@ -0,0 +1,587 @@
+
+We consider the first pillar to be more important, as theoretically it should be possible to provide simple
+APIs on top of exposed low-level internals. In this, we again are inspired by the Prosemirror model, where
+we see the community provide more structured wrappers (e.g. [Tiptap][tiptap]) on top of the internals, with more expressive but
+also more limited APIs.
+
+#### Immutability
+
+From past experience, we know that managing a state as complex as a document can be tough. While mutating state is easy to implement, 
+at various points in the application, we require detailed knowledge and control over those state changes. This is rather difficult to achieve when
+any piece of code can mutate the (in essence global) document state.
+
+We can learn from ProseMirror that a functional style lends itself well to these requirements. Treating the document as a value means things like
+combining transformations, history rollbacks, and asynchronous editing become simpler to reason about. Edits can become transactions, not unlike database
+transactions, which we like for their atomicity and ability to rollback to a known valid state when things go wrong.
+
+
+### Architecture Stage 1
+
+To give a quick overview of the ideas in this proposal, we will give an example of how a typical editing command would be implemented:
+
+The details and structuring of the various utility methods implied here are not part of this RFC, but they show a 
+philosophy of where various pieces of logic are expected to live.
+
+From the start of the VDOM-period, I've went back and forth about whether node classes
+should contain logic, or wether we should adopt a more functional design where 
+we make more heavy use of utility functions. From our experiments with a ModelTable
+custom node class, we learn that node-owned methods can be very expressive and intuitive.
+The question of "what can I do with this" becomes easily answered. 
+
+The main argument against this usually involved control over state changes. How
+do we implement a `table.insertRow` method while also making sure we can control and track any changes
+to the document? The current implementation of mutators and operations provides a solution, but
+we run into weird situations where we have methods on nodes which all take in a mutator, which they need to get from somewhere.
+It also puts a lot of strain on the mutator's api surface, as different node types can have different abstractions and might
+want to interact with their content in different ways.
+
+These problems dissapear if we apply the immutability principle here. If we demand that any node-owned method
+does not mutate the node itself (rather returns a new node with desired values), we can reduce the mutator api surface
+to convenient ways to replace existing nodes (or ranges) with their newly minted content. Prosemirror shows us that this
```

Comment:
I'm not really convinced by this, are you sure that prosemirror doesn't calculate the differences between the old node and the new node so it doesn't rerender it? I know React does this with the shadow dom so that may be the case with prosemirror also

Reply:
it does, that's what I mean that it can be implemented efficiently

Reply:
isn't that kinda hard to implement?